modules/dreamview/backend/teleop/teleop.cc

```diff
@@ -104,23 +149,70 @@ void TeleopService::RegisterMessageHandlers() {
       });
 }
 
-void TeleopService::UpdateModemInfo(
-    const std::shared_ptr<ModemInfo> &modem_info) {
-  // TODO simplify data and only send necessary info for display
-  if (modem_info->has_provider() && modem_info->has_technology()) {
-    boost::unique_lock<boost::shared_mutex> writer_lock(mutex_);
-    teleop_status_["modems"][modem_info->provider()] = modem_info->technology();
-  }
-}
 
 void TeleopService::SendStatus(WebSocketHandler::Connection *conn) {
   std::string to_send;
   {
     boost::shared_lock<boost::shared_mutex> reader_lock(mutex_);
     to_send = teleop_status_.dump();
+    AINFO << "status: " << to_send;
   }
   websocket_->SendData(conn, to_send);
 }
 
+
+void TeleopService::UpdateModem(const std::string &modem_id,
+    const std::shared_ptr<ModemInfo> &modem_info) {
+
+    // TODO simplify data and only send necessary info for display
+    // update modem_info_
+    if (modem_info->has_provider() && modem_info->has_technology()) {
+        // teleop_status_["modems"][modem_info->provider()] =
+        //  modem_info->technology();
+        boost::unique_lock<boost::shared_mutex> writer_lock(mutex_);
+        teleop_status_["modems"][modem_id] = modem_info->technology();
+    }
+}
+
+void TeleopService::UpdateCarDaemonRpt(
+    const std::shared_ptr<DaemonServiceRpt> &daemon_rpt) {
+  {
+      bool aVideoEncoderIsRunning = false;
+      bool voipIsRunning = false;
+      for (int i = 0; i < daemon_rpt->services_size(); i++) {
+          // look for voip_encoder or encoder0..1.2
+          // check 'voip_encoder' first because it contains 'encoder'
+          std::string service =  daemon_rpt->services(i);
+          if (service.find("voip_encoder") >= 0) {
+              voipIsRunning = true;
+          }
+          else if (service.find("encoder") >= 0) {
+              aVideoEncoderIsRunning = true; 
+          }
+
+      }
+      boost::unique_lock<boost::shared_mutex> writer_lock(mutex_);
+      teleop_status_["video"] = aVideoEncoderIsRunning;
+      teleop_status_["audio"] = voipIsRunning;
+  }
+}
+
+void TeleopService::UpdateOperatorDaemonRpt(
+    const std::shared_ptr<DaemonServiceRpt> &daemon_rpt) {
+  {
+      bool voipIsRunning = false;
+      for (int i = 0; i < daemon_rpt->services_size(); i++) {
+          std::string service =  daemon_rpt->services(i);
+          if (service.find("voip_encoder") >= 0) {
+              voipIsRunning = true;
```

Comment:
maybe break the for loop once voipIsRunning is set?

Reply:
added in this [commit](https://github.com/ApolloAuto/apollo/pull/9963/commits/84332ac9b452b2a739a1db47fb25f1568fa85827)